### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Zelinskyi-Serhii**
[DEMO LINK](https://Zelinskyi-Serhii.github.io/react_todo-app-with-api/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,24 +1,304 @@
...
+import { ErrorMessageType } from './types/ErrorMessageType';
 
-const USER_ID = 0;
+export const App: FC = () => {
+  const [allTodos, setAllTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState(ErrorMessageType.NONE);
+  const [todoTitle, setTodoTitle] = useState('');
+  const [filterType, setFilterType] = useState<FilterBy>(FilterBy.ALL);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [activeInput, setActiveInput] = useState(true);
```
Comment (illiarerun) [inline src/App.tsx:33]: do you need it? you can calculate it based on `tempTodo` value
```suggestion
```

File: src/App.tsx [hunk 2]
```diff
@@ -1,24 +1,304 @@
...
+  const countActiveTodos = filterTodosByCompleted(
+    allTodos,
+    FilterBy.ACTIVE,
+  ).length;
+
+  const isCompletedTodo = allTodos.length - countActiveTodos > 0;
+  const hasActiveTodo = countActiveTodos > 0;
+
+  const handleAddNewTodo = useCallback(
+    (event: React.FormEvent<HTMLFormElement>) => {
```
Comment (illiarerun) [inline src/App.tsx:117]: everything that works with form should be In corresponding Form component, you should just pass callback for adding a todo to it

File: src/App.tsx [hunk 3]
```diff
@@ -1,24 +1,304 @@
...
+          if (findTodo?.id === todo.id) {
+            findTodo.completed = !findTodo.completed;
+
+            return findTodo;
+          }
+
+          return todo;
+        });
+
+        setAllTodos(newTodos);
```
Comment (illiarerun) [inline src/App.tsx:165]: functional setState

File: src/App.tsx [hunk 4]
```diff
@@ -1,24 +1,304 @@
...
+        if (hasActiveTodo) {
+          if (!todo.completed) {
+            await setLoadingTodoIds(prev => [...prev, todo.id]);
+          }
+        } else if (todo.completed) {
+          await setLoadingTodoIds(prev => [...prev, todo.id]);
+        }
+
+        return updateTodoCompleted(todo.id, hasActiveTodo);
+      });
```
Comment (illiarerun) [inline src/App.tsx:186]: put it inside try/catch cause you call `update` here

File: src/App.tsx [hunk 5]
```diff
@@ -1,24 +1,304 @@
...
+
+            updatedTodo.title = newTitle;
+
+            return updatedTodo;
+          }
+
+          return todo;
+        });
+
+        setAllTodos(newTodos);
```
Comment (illiarerun) [inline src/App.tsx:224]: functional setState

File: src/App.tsx [hunk 6]
```diff
@@ -1,24 +1,304 @@
...
+            )}
+          />
+
+          <Form
+            handleAddTodo={handleAddNewTodo}
+            todoTitle={todoTitle}
+            setTodoTitle={setTodoTitle}
+            activeInput={activeInput}
+          />
+        </header>
```
Comment (illiarerun) [inline src/App.tsx:257]: Header component

File: src/App.tsx [hunk 7]
```diff
@@ -1,24 +1,304 @@
...
+
+        {(!allTodos.length && isRequest && !errorMessage)
```
Comment (illiarerun) [inline src/App.tsx:259]: this would be enough
```suggestion
        {(!allTodos.length
```

File: src/App.tsx [hunk 8]
```diff
@@ -1,24 +1,304 @@
...
+            display: 'flex',
+            justifyContent: 'center',
+            color: 'grey',
+          }}
+          >
+            You have not any todos
+          </span>
+        )}
+
+        {!allTodos.length && !errorMessage && !isRequest
```
Comment (illiarerun) [inline src/App.tsx:271]: would be enough
```suggestion
        {!allTodos.length && !errorMessage 
```

File: src/App.tsx [hunk 9]
```diff
@@ -1,24 +1,304 @@
...
+                tempTodos={tempTodo}
+                handleRemoveTodo={handleRemoveTodo}
+                loadingTodoIds={loadingTodoIds}
+                handleUpdateTodoCompleted={handleUpdateTodoCompleted}
+                changeTitleByDoubleClick={changeTitleByDoubleClick}
+              />
+            </section>
+          )}
+
+        {!allTodos.length || (
```
Comment (illiarerun) [inline src/App.tsx:287]: why don't just use `&&`?
```suggestion
        {allTodos.length && (
```

File: src/components/TodoList/TodoList.tsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+              key={todo.id}
+              handleRemoveTodo={handleRemoveTodo}
+              loadingTodoIds={loadingTodoIds}
+              handleUpdateTodoCompleted={handleUpdateTodoCompleted}
+              changeTitleByDoubleClick={changeTitleByDoubleClick}
+            />
+          </CSSTransition>
+        ))}
+
+        {tempTodos && (
```
Comment (illiarerun) [inline src/components/TodoList/TodoList.tsx:47]: it's only one todo
```suggestion
        {tempTodo && (
```

**Review by illiarerun — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 11]
```diff
@@ -1,24 +1,293 @@
...
 
-const USER_ID = 0;
+export const App: FC = () => {
+  const [allTodos, setAllTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState(ErrorMessageType.NONE);
+  const [todoTitle, setTodoTitle] = useState('');
+  const [filterType, setFilterType] = useState<FilterBy>(FilterBy.ALL);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isLoadingTodos, setIsLoadingTodos] = useState(false);
+  const [loadingIds, setLoadingIds] = useState<Set<number>>(new Set());
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:33]: interesting approach

File: src/App.tsx [hunk 12]
```diff
@@ -1,24 +1,293 @@
...
+  const loadAllTodos = async () => {
+    setIsLoadingTodos(true);
+    try {
+      const todos = await getTodos(USER_ID);
+
+      setAllTodos(todos);
+    } catch {
+      showError(ErrorMessageType.LOAD);
+    } finally {
+      setTempTodo(null);
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:73]: Strange that you do it here, 'cause you did it in `handleAdd` method

File: src/App.tsx [hunk 13]
```diff
@@ -1,24 +1,293 @@
...
-const USER_ID = 0;
+export const App: FC = () => {
+  const [allTodos, setAllTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState(ErrorMessageType.NONE);
+  const [todoTitle, setTodoTitle] = useState('');
+  const [filterType, setFilterType] = useState<FilterBy>(FilterBy.ALL);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isLoadingTodos, setIsLoadingTodos] = useState(false);
+  const [loadingIds, setLoadingIds] = useState<Set<number>>(new Set());
+  const addLoadingTodo = useCallback((id: number) => {
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:34]:
```suggestion
  const setTodoLoading = useCallback((id: number) => {
```

File: src/App.tsx [hunk 14]
```diff
@@ -1,24 +1,293 @@
...
+    const completedTodos = allTodos.filter(({ completed }) => completed);
+
+    completedTodos.forEach(async todo => {
+      try {
+        addLoadingTodo(todo.id);
+        await removeTodo(todo.id);
+
+        setAllTodos(prevTodos => prevTodos.filter(({ completed }) => (
+          !completed
+        )));
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:109]: Should we filter array on every iteration? I guess better to move it outside forEach, or reuse existing `handleRemoveTodo` method here

File: src/App.tsx [hunk 15]
```diff
@@ -1,24 +1,293 @@
...
+      } catch {
+        setErrorMessage(ErrorMessageType.DELETE);
+      } finally {
+        setLoadingIds(new Set());
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:113]: Same here, as soon as first todo will be removed - state will be cleared which is not logical

Comment (OleksiiBlezniuk) [reply src/App.tsx:113]: Or reuse existing method for consistency, in any case duplicating logic with not logical changes is bad idea

File: src/App.tsx [hunk 16]
```diff
@@ -1,24 +1,293 @@
...
+  const visibleTodos = useMemo(() => {
+    return filterTodosByCompleted(allTodos, filterType);
+  }, [allTodos, filterType]);
+
+  const countActiveTodos = useMemo(() => (
+    filterTodosByCompleted(
+      allTodos,
+      FilterBy.ACTIVE,
+    )
+  ), [allTodos, filterType]).length;
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:127]: Looks strange, let's save result of filter to constant and then save it length to separate constant

File: src/App.tsx [hunk 17]
```diff
@@ -1,24 +1,293 @@
...
+
+  const isCompletedTodo = allTodos.length - countActiveTodos > 0;
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:129]: which todo is completed? use more clear names

Comment (OleksiiBlezniuk) [inline src/App.tsx:129]:
```suggestion
  const isAnyTodoIsCompleted = allTodos.length - countActiveTodos > 0;
```

File: src/App.tsx [hunk 18]
```diff
@@ -1,24 +1,293 @@
...
+        setTodoTitle('');
+        setTempTodo(null);
+      }
+    }, [todoTitle],
+  );
+
+  const handleUpdateTodoCompleted = useCallback(
+    async (todoId: number) => {
+      try {
+        const findTodo = allTodos.find(todo => todo.id === todoId);
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:163]: do it in setState, and use more clear name for it

File: src/App.tsx [hunk 19]
```diff
@@ -1,24 +1,293 @@
...
+    async () => {
+      try {
+        const allTodosPromises = allTodos.map(async (todo) => {
+          if (hasActiveTodo) {
+            if (!todo.completed) {
+              addLoadingTodo(todo.id);
+            }
+          } else if (todo.completed) {
+            addLoadingTodo(todo.id);
+          }
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:197]: Can we simplify this check? Such nested ifs looks complicated

File: src/components/TododInfo/TodoInfo.tsx [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+              className="todoapp__new-todo"
+              style={{
+                paddingLeft: 15,
+              }}
+              placeholder="Empty todo will be deleted"
+              value={newTitle}
+              onChange={event => setNewTitle(event.target.value)}
+            />
+
+          </form>
```
Comment (OleksiiBlezniuk) [inline src/components/TododInfo/TodoInfo.tsx:105]: Would be nice to move it to separate component, but it's not critical

File: src/App.tsx [hunk 21]
```diff
@@ -1,24 +1,285 @@
...
+            setTodoLoading(todo.id);
+          }
+
+          if (!hasActiveTodo && todo.completed) {
+            setTodoLoading(todo.id);
+          }
+
+          return updateTodoCompleted(todo.id, hasActiveTodo);
+        });
+        const updatedTodos = await Promise.all(allTodosPromises);
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:193]:
```suggestion
        const updatedTodos = await Promise.all(
          allTodos.map(async (todo) => {
            if (hasActiveTodo && !todo.completed) {
              setTodoLoading(todo.id);
            }
  
            if (!hasActiveTodo && todo.completed) {
              setTodoLoading(todo.id);
            }
  
            return updateTodoCompleted(todo.id, hasActiveTodo);
          })
        );
```

File: src/components/ErrorMessage/ErrorMessage.tsx [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+  errorMessage: string;
+  setErrorMessage: (message: ErrorMessageType) => void;
+};
+
+export const ErrorMessage: React.FC<Props> = React.memo(
+  ({
+    errorMessage,
+    setErrorMessage,
+  }) => {
+    return (
```
Comment (OleksiiBlezniuk) [inline src/components/ErrorMessage/ErrorMessage.tsx:15]: I don't think this `return` keyword is needed here

File: src/components/Form/Form.tsx [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+        <input
+          type="text"
+          disabled={!activeInput}
+          className="todoapp__new-todo"
+          placeholder="What needs to be done?"
+          value={todoTitle}
+          onChange={({ target }) => setTodoTitle(target.value)}
+        />
+
+      </form>
```
Comment (OleksiiBlezniuk) [inline src/components/Form/Form.tsx:34]:
```suggestion
        />
      </form>
```

File: src/components/Header/Header.tsx [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+
+export const Header: React.FC<Props> = ({
+  handleChangeStatusAllTodos,
+  hasActiveTodo,
+  handleAddNewTodo,
+  todoTitle,
+  setTodoTitle,
+  activeInput,
+
+}) => (
```
Comment (OleksiiBlezniuk) [inline src/components/Header/Header.tsx:22]:
```suggestion
  todoTitle,
  setTodoTitle,
  activeInput,
}) => (
```

**Review by OleksiiBlezniuk — APPROVED** [review]

**Pull request closed without merging**